src/platform_impl/android/mod.rs

```diff
@@ -1,171 +1,240 @@
+        v.push_back(self.primary_monitor());
+        v
+    }
+
+    pub fn create_proxy(&self) -> EventLoopProxy<T> {
+        EventLoopProxy {
+            tx: self.window_target().p.tx.clone(),
+            queue: self.user_queue.clone(),
+        }
+    }
+
+    fn glue_event_to_event(&self, glue_event: android_glue::Event) -> Option<event::Event<T>> {
+        let window_id = crate::window::WindowId(WindowId);
+        let device_id = event::DeviceId(DeviceId);
+
+        match glue_event {
+            android_glue::Event::EventMotion(motion_event) => Some(event::Event::WindowEvent {
+                window_id,
+                event: event::WindowEvent::Touch(event::Touch {
+                    device_id,
+                    phase: match motion_event.action {
+                        android_glue::MotionAction::Down => event::TouchPhase::Started,
+                        android_glue::MotionAction::Up => event::TouchPhase::Ended,
+                        android_glue::MotionAction::Move => event::TouchPhase::Moved,
+                        android_glue::MotionAction::Cancel => event::TouchPhase::Cancelled,
+                    },
+                    location: crate::dpi::LogicalPosition::from_physical(
+                        (motion_event.x as f64, motion_event.y as f64),
+                        self.get_dpi_factor(),
+                    ),
+                    id: motion_event.pointer_id as u64,
+                }),
+            }),
+            android_glue::Event::User => Some(event::Event::UserEvent(
+                self.user_queue.lock().unwrap().pop_front().unwrap(),
+            )),
+            android_glue::Event::Start => None,
+            android_glue::Event::Pause => None,
+            android_glue::Event::Resume => None,
+            android_glue::Event::Stop => None,
+            android_glue::Event::Destroy => Some(event::Event::LoopDestroyed),
+            android_glue::Event::SaveState => None, // TODO save the state now
+            android_glue::Event::LostFocus => Some(event::Event::WindowEvent {
+                window_id: window::WindowId(WindowId),
+                event: event::WindowEvent::Focused(false),
+            }),
+            android_glue::Event::GainedFocus => Some(event::Event::WindowEvent {
+                window_id: window::WindowId(WindowId),
+                event: event::WindowEvent::Focused(true),
+            }),
+            android_glue::Event::ConfigChanged => None, // TODO maybe notify HiDPI changed and/or resized
+            _ => None,
         }
     }
 
-    pub fn create_proxy(&self) -> EventLoopProxy {
-        EventLoopProxy
+    // FIXME this is not true
+    fn get_dpi_factor(&self) -> f64 {
+        1.
     }
 }
 
-impl EventLoopProxy {
-    pub fn wakeup(&self) -> Result<(), ::EventLoopClosed> {
-        android_glue::wake_event_loop();
-        Ok(())
+#[derive(Clone)]
+pub struct EventLoopProxy<T: 'static> {
+    tx: mpsc::Sender<android_glue::Event>,
```

Comment:
Can't this be `mpsc::Sender<Option<android_glue::Event>>` instead, so we can send `None` instead of `android_glue::Event::User`?

Reply:
`android-rs-glue` will need to be changed either way; it's the one that produces events on the channel.

Reply:
We are the only ones sending the `User` event, right here: https://github.com/rust-windowing/winit/pull/1001/files/f081128c792f6ab52511294327845deda0a4ca9a#diff-2159684293d0c0086e6a3590e44ccba7R223

If we send `None` instead of the `User` event, then we don't need to change android-rs-glue.

Is there another place that is sending `android_glue::Event::User`? I know for sure that your PR to `android-rs-glue` isn't, cause it's only a 4 line change.

Reply:
You're right, winit is the only one using `android_glue::Event::User`.  However, the easiest way to nicely mesh user events with the events [already being sent](https://github.com/rust-windowing/android-rs-glue/blob/master/cargo-apk/injected-glue/lib.rs#L412) on the channel by `android-rs-glue` is to add it to the channel, and you can either add it to `android_glue::Event` or modify the channel `android-rs-glue` is using to deal in `Option<Event>`'s.  Either way, it's a change to `android-rs-glue`.